backend/src/impact.rs

```diff
@@ -0,0 +1,126 @@
+use std::collections::HashMap;
+
+use geojson::{Feature, FeatureCollection};
+use nanorand::{Rng, WyRand};
+
+use crate::{IntersectionID, MapModel, RoadID};
+
+// TODO Rename?
+/// Besides just studying the impact on shortcuts within one neighbourhood boundary, the user can
+/// see how traffic changes across roads in the whole map. This works by finding the best route
+/// before and after changes for every origin/destination "OD" pairs, then counting routes per
+/// road.
+pub struct Impact {
+    requests: Vec<(IntersectionID, IntersectionID)>,
+
+    // TODO Can use Vec for perf
+    counts_before: HashMap<RoadID, usize>,
+    counts_after: HashMap<RoadID, usize>,
+}
+
+impl Impact {
+    /// Calculates `requests` only
+    pub fn new(map: &MapModel) -> Self {
+        Self {
+            requests: synthetic_od_requests(map),
+            counts_before: HashMap::new(),
+            counts_after: HashMap::new(),
+        }
+    }
+
+    pub fn invalidate_after_edits(&mut self) {
+        self.counts_after.clear();
+    }
+
+    /// Returns a feature per road, with `before` and `after` counts, and a `max_count` foreign
+    /// member
+    pub fn recalculate(&mut self, map: &MapModel) -> FeatureCollection {
+        if self.counts_before.is_empty() {
+            info!("Calculating impacts before edits");
+            self.counts_before = map
+                .router_before
+                .as_ref()
+                .unwrap()
+                .od_to_counts(map, &self.requests);
+        }
+
+        if self.counts_after.is_empty() {
+            info!("Calculating impacts after edits");
+            self.counts_after = map
+                .router_after
+                .as_ref()
+                .expect("need to rebuild_router")
+                .od_to_counts(map, &self.requests);
+        }
+
+        let mut features = Vec::new();
+        let mut max_count = 0;
+        for road in &map.roads {
+            let before = self.counts_before.get(&road.id).cloned().unwrap_or(0);
+            let after = self.counts_after.get(&road.id).cloned().unwrap_or(0);
+            // Don't show unchanged roads, but to scale the absolute counts, do look at the max
+            // count seen anywhere
+            max_count = max_count.max(before.max(after));
+            if before != after && (before > 0 || after > 0) {
+                let mut f = map.mercator.to_wgs84_gj(&road.linestring);
+                f.set_property("id", road.id.0);
+                f.set_property("before", before);
+                f.set_property("after", after);
+                features.push(f);
+            }
+        }
+
+        FeatureCollection {
+            features,
+            bbox: None,
+            foreign_members: Some(
+                serde_json::json!({
+                    "max_count": max_count,
+                })
+                .as_object()
+                .unwrap()
+                .clone(),
```

Comment:
minor nit: is this clone necessary - since it's the last thing in the method, I'd expect a move should be sufficient.

Reply:
Sadly yes, [as_object](https://docs.rs/serde_json/latest/serde_json/enum.Value.html#method.as_object) returns a ref, and there's no `into_object` helper